### ISSUE TITLE
[VoIP] Calls don't seem to work between devices on the same network?

### ISSUE THREAD

**Issue by opusforlife2** [vector-im/element-android#3389]
#### Describe the bug

The devices are on the same WiFi network.

Element-Android to Element-Web or vice versa: Call gets dropped automatically after trying to connect for a few seconds.

Element-Android to Element-Android: Keeps showing "Call connecting" indefinitely, though notifications start appearing in the background which notify of a dropped call.

Edit: One of the notifications on the Android device is 'Debug: event type "m.call.negotiate" '.

#### To Reproduce
1. Call from Device1, or through Element-Web.
2. Pick up on Device2.
3. Call either disconnects after trying for a few seconds, or keeps trying to connect indefinitely.

#### Expected behavior

Calls should work seamlessly between devices on the same WiFi network.

#### Smartphone (please complete the following information):
 - Device: Xiaomi Mi 5, Sony Xperia X Compact
 - OS: Android 10 (LOS 17.1), Android 8.0

#### Additional context
 - App version and store: 1.1.7 - F-Droid
 - Homeserver: fairydust.space, matrix.org

**Comment by opusforlife2** [comment]
Update: Weird new situation with this bug.

3 devices, all on the same WLAN. One is a Mi 5 (on fairydust.space), and the others are a Nokia 8 and Nokia 6.1 (both on matrix.org).
1:1 voice calls work between the Mi 5 and the Nokia 6.1, but not between the Mi 5 and the Nokia 8. It keeps showing "Call connecting" until timeout.

There is literally no difference between the app config between these devices.  Just the device models themselves are different. The situation remains the same when not on the same WLAN, or using mobile data. One pair of devices can successfully call each other, the other pair can't.

Element-Android v1.2.2 on all 3 devices.

@ganfra Could you also have a look at my other VoIP issues once? https://github.com/vector-im/element-android/issues?q=is%3Aopen+is%3Aissue+author%3Aopusforlife2+voip

**Comment by opusforlife2** [comment]
@ouchadam Hi, sorry for pinging you directly. Could you please help me debug this? I can't VoIP call a lot of my contacts due to this, including, as I've said above, family members. I've sent several rageshakes, since this problem occurs regularly, but nothing seems to have come out of them so far.

**Comment by ouchadam** [comment]
no problem :smile:  

it sounds like something fishy is going on either at the android OS level, potentially homeserver to homeserver or fdroid background sync issues

- are all devices running the f-droid variant?
- does voip work when the element app is open and in the foreground?
- is the problem intermittent or does it always fail for some device combinations?
- do calls work when reversing the call initiator?

**Comment by opusforlife2** [comment]
Thank you!

- No, mine is F-Droid, other two are Play Store (1 with working calls, 1 with not)
- Nope. Both ways. I call and it keeps showing "call connecting". The other person calls and it's the same. But it works with the other device.
- Always. Me with person 1 = always succeeds in calling. Me with person 2 = always stuck on call connecting.
- Nope.

**Comment by opusforlife2** [comment]
Oh, I've already explained it in detail in this comment: https://github.com/vector-im/element-android/issues/3389#issuecomment-922214337

**Comment by ouchadam** [comment]
what about element web to element web? (if possible)

**Comment by opusforlife2** [comment]
I can try that and get back to you.

**Comment by ouchadam** [comment]
thanks! I'll try and grab some time to see if I can reproduce on my side

**Comment by opusforlife2** [comment]
Could things like DoT/DoH config interfere with the call setup process?

**Comment by ouchadam** [comment]
I would hope not as we're using external servers for the voice transport :see_no_evil: 

something else to double check is if `matrix.org` to `matrix.org` accounts behave differently on the same devices (to help rule out as many options as possible!)

**Comment by opusforlife2** [comment]
@ouchadam Tried it with this config just now:

- Device: Xiaomi Mi 5, Sony Xperia X Compact
- OS: Android 10 (LOS 17.1), Android 8.0

Logged in to Element Web from the accounts on both these devices:

- Call placed from either account.
- Rings on other account.
- Pick up.
- Call drops within 2-3 seconds.
- Timeline message says "Connection failed".

**Comment by opusforlife2** [comment]
@ouchadam?

**Comment by ouchadam** [comment]
thanks for the update, this helps narrow things down a lot (from a client perspective at least)!

could you try running https://test.webrtc.org/ 

I'm leaning towards network configuration causing problems, potentially blocking the webrtc negotiation

**Comment by opusforlife2** [comment]
Thanks for that test link! It helped a lot in debugging. I also tested using Jitsi Meet.

> Could things like DoT/DoH config interfere with the call setup process?

My hunch was right. It turns out that if DoT is enabled natively using the option in Android 9+, then everything works fine.

But if you use an app to enforce DoT/DoH, then it breaks WebRTC connections (tested with Jitsi Meet as well). Allowlisting Element in the app fixed the problem. (Strangely, video still didn't work in Jitsi Meet after allowlisting.)

So Android <-> Android calls have been fixed. Android <-> Web works as well. Web <-> Web still doesn't work, but that can't have anything to do with the system DNS resolver, because Firefox does its own thing there.

For my purposes, this issue is solved, so I'm closing it. If I figure out how to get Web <-> Web calls working, I'll leave a comment here.

Thank you so much, @ouchadam!
